File: inheritance-and-polymorphism/code/institute.cpp
#include <iostream>
#include <string>

using namespace std;
class Professor {
protected:
  const string name;
public:
  Professor(string n) : name(n) {}
  
  void teach(string course, string cl) {
    cout << "Prof. " << name << " teaches course " << course
      << " to class " << cl << "." << endl;
  }
};

class Warden {
protected:
  const string name;
public:
  Warden(string n) : name(n) {}
  void approve(string request) {
    cout << "Warden " << name << " approves request " << request << "." << endl;
  }
};

class ProfessorWarden : public Professor, public Warden {
public:
  ProfessorWarden(string n) : Professor(n), Warden("ditto") {}
  void doWork() {
    cout << "Prof. " << Professor::name << " teaches and approves requests." << endl;
  }
};

int main() {
  Professor prof("Sujit");
  prof.teach("Java", "MT2017");

  Warden warden("Sujit");
  warden.approve("Dinner out");

  ProfessorWarden profwarden("SKC");
  
  Professor *profptr = &profwarden;
  Warden * wardenptr = &profwarden;

  profptr->teach("Python", "IMT2017");
  wardenptr->approve("party request");

  ProfessorWarden * profwardenptr = &profwarden;
  profwardenptr->teach("Programming Languages", "IMT2013");
  profwardenptr->approve("Cultural festival request");
  profwardenptr->doWork();

  return 0;
}

File: inheritance-and-polymorphism/code/Institute.java
public class Institute {
  public static void main(String[] a) {

    ProfessorWarden pw = new ProfessorWarden("SKC");

    IProfessor p = pw;
    IWarden    w = pw;

    p.teach("OOP", "Upgrad");
    w.approve("Festival request");

    pw.teach("OOP", "Upgrad");
    pw.approve("Party request");
    pw.doWork();
  }
}

interface IProfessor {
  void teach(String course, String cl);
}

interface IWarden {
  void approve(String request);
}

class Person {
  protected String name;

  public Person(String n) {
    this.name = n;
  }
}

class Professor extends Person implements IProfessor {
  public Professor(String n) {
    super(n);
  }

  public void teach(String course, String cl) {
    System.out.println("Prof. " + this.name + " teaches " + course + " to " + cl + ".");
  }
}

class Warden extends Person implements IWarden {
  public Warden(String n) {
    super(n);
  }

  public void approve(String request) {
    System.out.println("Warden " + this.name + " approves request " + request + ".");
  }
}

class ProfessorWarden extends Person implements IProfessor, IWarden {
   public ProfessorWarden(String n) {
    super(n);
  }

  public void teach(String cl, String s) {
    System.out.println("Prof. " + this.name + " teaches " + s + " to " + cl + ".");
  }

  public void approve(String request) {
    System.out.println(this.name + "  " + request + ".");
  }

  public void doWork() {
    System.out.println("Prof. " + this.name + " teaches and approves requests.");
  }
}
